### ISSUE TITLE
Exact object types shouldn't allow array access without an indexer annotation

### ISSUE THREAD

**Issue by popham** [facebook/flow#4257]
Flow passes the following cleanly, but it shouldn't:

```js
type T = {||};
declare var k: T;
k[1] = 7;
```

Similar to https://github.com/facebook/flow/issues/3162.

**Comment by jfirebaugh** [comment]
I'd expect even a _non_-exact object type to fail with the "access of computed property/element . Indexable signature not found" error. The following should be a type error:

```
type Foo = { n: number };
const foo: Foo = {n: 0};
const name: string = 'n';
const prop: boolean = foo[name];
```

I want computed property access to be allowed only when there's an indexer property.

**Comment by jcready** [comment]
@jfirebaugh if you remove the `string` annotation from `name` it will give you your flow error.

```js
type Foo = { n: number };
const foo: Foo = {n: 0};
const name = 'n';
const prop: boolean = foo[name];
```

**Comment by jfirebaugh** [comment]
Interesting -- I suppose in that case it gets typed as a literal, so essentially equivalent to `foo['n']` which is also an error.

In the context where I encountered the issue, that's not an option however; `name` is not assigned from a string literal, and not explicitly annotated -- flow infers a `string` type.

**Comment by jcready** [comment]
Well, I believe [this is the source](https://github.com/facebook/flow/blob/master/lib/core.js#L59) of your problem. You can again work around this by changing your type definition:

```js
type Foo = { n: number, [any]: mixed };
```

**Comment by roberthoenig** [comment]
Bump.

**Comment by SamChou19815** [comment]
Fixed in 0.190.0
